File: source/sdk/go/expressions/value.go
package expressions

import (
	"fmt"
	"math"
	"sort"
	"strconv"
	"strings"
	"time"
)

// The value space.
//
// Two rules make everything else in this package tractable, and both are pinned
// by ADR-0009:
//
//   - There is exactly one absent value, nil. A missing key yields it, and
//     reading a property of it yields it again.
//   - There is one numeric type and it is float64, so `7 / 2` is 3.5 here as
//     well as in TypeScript. Never reach for int.
//
// NaN and Inf are not in the space at all: division by zero is an error rather
// than a value, which is what keeps them out.

// ValueType is every type a value or a declaration can name. The first seven
// are the Component Manifest's own output types, so a field's declared type and
// an expression's type are drawn from one vocabulary rather than two that have
// to be mapped. TypeUnknown and TypeNull exist only in the language.
type ValueType string

const (
	TypeText     ValueType = "text"
	TypeNumber   ValueType = "number"
	TypeBoolean  ValueType = "boolean"
	TypeDatetime ValueType = "datetime"
	TypeList     ValueType = "list"
	TypeObject   ValueType = "object"
	TypeItem     ValueType = "item"
	TypeUnknown  ValueType = "unknown"
	TypeNull     ValueType = "null"
)

// Value is string, float64, bool, time.Time, []Value, map[string]Value or nil.
type Value = any

// TypeOf reports a value's type in the same vocabulary a manifest declares.
//
// The integer cases are not part of the value space — there is one numeric type
// and it is float64 — but they are what a YAML decoder produces, including this
// SDK's own LoadDefinition. Recognising them here means a Host that feeds
// decoded YAML straight in gets a number rather than being told its count is an
// object; Normalize is what actually converts them.
func TypeOf(value Value) ValueType {
	switch value.(type) {
	case nil:
		return TypeNull
	case string:
		return TypeText
	case float64, float32, int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64:
		return TypeNumber
	case bool:
		return TypeBoolean
	case time.Time:
		return TypeDatetime
	case []Value:
		return TypeList
	case map[string]Value:
		return TypeObject
	}
	return TypeObject
}

// Normalize converts a value into the value space: every number becomes a
// float64, and lists and objects are converted through.
//
// It exists because the value space and Go's YAML and JSON decoders disagree
// about integers. `count: 24` decodes to an `int`, and an `int` reaching the
// evaluator used to be typed `object` and rendered as "null" — a silent wrong
// answer rather than a loud failure. Values arriving from a Host are normalized
// as they are read, so a runner can hand over decoded YAML without knowing any
// of this.
func Normalize(value Value) Value {
	switch v := value.(type) {
	case int:
		return float64(v)
	case int8:
		return float64(v)
	case int16:
		return float64(v)
	case int32:
		return float64(v)
	case int64:
		return float64(v)
	case uint:
		return float64(v)
	case uint8:
		return float64(v)
	case uint16:
		return float64(v)
	case uint32:
		return float64(v)
	case uint64:
		return float64(v)
	case float32:
		return float64(v)
	case []any:
		out := make([]Value, 0, len(v))
		for _, item := range v {
			out = append(out, Normalize(item))
		}
		return out
	case map[string]any:
		out := make(map[string]Value, len(v))
		for key, item := range v {
			out[key] = Normalize(item)
		}
		return out
	}
	return value
}

// Satisfies reports whether a value matches a declared type.
//
// Coercion at the boundary is narrow and declared, so "must match" has a
// precise meaning: any scalar into text is permitted, because a text field is
// the universal sink and that is exactly what interpolation already does; text
// into number is not implicit and requires num.parse(); nil satisfies any
// declared type, since whether absence is acceptable is `req:`'s business and
// not the evaluator's; everything else must match exactly.
func Satisfies(value Value, declared ValueType) bool {
	if declared == TypeUnknown || declared == TypeItem {
		return true
	}
	if value == nil {
		return true
	}
	actual := TypeOf(value)
	if actual == declared {
		return true
	}
	if declared == TypeText {
		return IsScalar(actual)
	}
	return false
}

// IsScalar reports whether a type is one of the four that can render as text.
func IsScalar(t ValueType) bool {
	return t == TypeText || t == TypeNumber || t == TypeBoolean || t == TypeDatetime
}

// AsText renders a value at a text boundary.
func AsText(value Value) string {
	switch v := value.(type) {
	case nil:
		return ""
	case string:
		return v
	case float64:
		return NumberToText(v)
	case bool:
		if v {
			return "true"
		}
		return "false"
	case time.Time:
		return DatetimeToText(v)
	}
	return ToJSON(value)
}

// ToJSON is canonical JSON — what json.stringify produces.
//
// Hand-rolled in both languages rather than reaching for the built-in, because
// the two built-ins disagree in ways that would reach a user: Go sorts object
// keys and JavaScript preserves insertion order, Go escapes `<` and `&` by
// default, and each formats numbers its own way. Sorting keys in both is the
// only choice that can be made identical, so both sort.
func ToJSON(value Value) string {
	switch v := value.(type) {
	case nil:
		return "null"
	case string:
		return jsonString(v)
	case float64:
		return NumberToText(v)
	case bool:
		if v {
			return "true"
		}
		return "false"
	case time.Time:
		return jsonString(DatetimeToText(v))
	case []Value:
		parts := make([]string, 0, len(v))
		for _, item := range v {
			parts = append(parts, ToJSON(item))
		}
		return "[" + strings.Join(parts, ",") + "]"
	case map[string]Value:
		keys := make([]string, 0, len(v))
		for key := range v {
			keys = append(keys, key)
		}
		sort.Strings(keys)
		parts := make([]string, 0, len(keys))
		for _, key := range keys {
			parts = append(parts, jsonString(key)+":"+ToJSON(v[key]))
		}
		return "{" + strings.Join(parts, ",") + "}"
	}
	return "null"
}

func jsonString(value string) string {
	var out strings.Builder
	out.WriteByte('"')
	for _, char := range value {
		switch char {
		case '"':
			out.WriteString(`\"`)
		case '\\':
			out.WriteString(`\\`)
		case '\n':
			out.WriteString(`\n`)
		case '\r':
			out.WriteString(`\r`)
		case '\t':
			out.WriteString(`\t`)
		case '\b':
			out.WriteString(`\b`)
		case '\f':
			out.WriteString(`\f`)
		default:
			if char < 0x20 {
				out.WriteString(fmt.Sprintf(`\u%04x`, char))
				continue
			}
			out.WriteRune(char)
		}
	}
	out.WriteByte('"')
	return out.String()
}

// DatetimeToText renders an instant as RFC 3339 in UTC, with a fractional part
// only when there is one.
//
// RFC3339Nano is exactly that; the note is for the TypeScript side, where
// Date.toISOString() always writes three decimal places and has to be trimmed
// to match this.
//
// Truncated to milliseconds here rather than only in dt.parse and dt.now,
// because a Host puts instants into the Context too — time.Now() carries
// nanoseconds, and a Date cannot. Doing it at the rendering boundary is what
// makes "instants carry millisecond precision" true of every instant rather
// than only of the ones Hatua created.
func DatetimeToText(value time.Time) string {
	return value.UTC().Truncate(time.Millisecond).Format(time.RFC3339Nano)
}

// NumberToText is the ECMAScript Number::toString algorithm, ported.
//
// Go's own formatting disagrees with JavaScript's at both ends of the range:
// strconv gives "1e-06" where JavaScript gives "0.000001". A workflow that
// emails a number must not read differently depending on which runner sent it,
// so the JavaScript spelling wins and this is what makes Go produce it.
//
// The shortest round-trip digits come from strconv with precision -1, which is
// exactly the "k as small as possible" the specification asks for.
func NumberToText(value float64) string {
	if math.IsNaN(value) {
		return "NaN"
	}
	if math.IsInf(value, 1) {
		return "Infinity"
	}
	if math.IsInf(value, -1) {
		return "-Infinity"
	}
	if value == 0 {
		// Covers -0, which JavaScript renders as "0".
		return "0"
	}
	if value < 0 {
		return "-" + NumberToText(-value)
	}

	// `d.dddde±dd` — the mantissa digits are s, and the exponent is n-1.
	scientific := strconv.FormatFloat(value, 'e', -1, 64)
	mantissa, exponent, _ := strings.Cut(scientific, "e")
	digits := strings.Replace(mantissa, ".", "", 1)
	power, err := strconv.Atoi(exponent)
	if err != nil {
		return scientific
	}

	k := len(digits)
	n := power + 1

	switch {
	case k <= n && n <= 21:
		return digits + strings.Repeat("0", n-k)
	case 0 < n && n <= 21:
		return digits[:n] + "." + digits[n:]
	case -6 < n && n <= 0:
		return "0." + strings.Repeat("0", -n) + digits
	}

	sign := "+"
	e := n - 1
	if e < 0 {
		sign = "-"
		e = -e
	}
	if k == 1 {
		return digits + "e" + sign + strconv.Itoa(e)
	}
	return digits[:1] + "." + digits[1:] + "e" + sign + strconv.Itoa(e)
}

File: source/packages/expressions/src/value.ts
/**
 * The value space.
 *
 * Two rules make everything else in this package tractable, and both are
 * pinned by ADR-0009:
 *
 *   - There is exactly one absent value, `null`. A missing key yields it, and
 *     reading a property of it yields it again. Nothing here ever produces
 *     `undefined`.
 *   - There is one numeric type and it is a 64-bit float, so `7 / 2` is 3.5 in
 *     both languages. Go must never reach for `int`.
 *
 * `NaN` and `Infinity` are not in the space at all: division by zero is an
 * error rather than a value, which is what keeps them out.
 */

/**
 * Every type a value or a declaration can name.
 *
 * The first seven are the Component Manifest's own output types, so a field's
 * declared type and an expression's type are drawn from one vocabulary rather
 * than two that have to be mapped. `unknown` and `null` exist only here.
 */
export type ValueType =
  | 'text'
  | 'number'
  | 'boolean'
  | 'datetime'
  | 'list'
  | 'object'
  | 'item'
  | 'unknown'
  | 'null'

export type Value =
  | string
  | number
  | boolean
  | Date
  | readonly Value[]
  | { readonly [key: string]: Value }
  | null

/** The runtime type of a value, in the same vocabulary a manifest declares. */
export function typeOf(value: Value): ValueType {
  if (value === null) return 'null'
  if (typeof value === 'string') return 'text'
  if (typeof value === 'number') return 'number'
  if (typeof value === 'boolean') return 'boolean'
  if (value instanceof Date) return 'datetime'
  if (Array.isArray(value)) return 'list'
  return 'object'
}

/**
 * Whether a value satisfies a declared type.
 *
 * Coercion at the boundary is narrow and declared, so "must match" has a
 * precise meaning:
 *
 *   - any scalar into `text` is permitted — a `text` field is the universal
 *     sink, and that is exactly what interpolation already does;
 *   - `text` into `number` is *not* implicit; it requires `num.parse()`;
 *   - `null` satisfies any declared type. Whether absence is *acceptable* is
 *     `req:`'s business, not the evaluator's;
 *   - everything else must match exactly.
 */
export function satisfies(value: Value, declared: ValueType): boolean {
  if (declared === 'unknown' || declared === 'item') return true
  if (value === null) return true

  const actual = typeOf(value)
  if (actual === declared) return true
  if (declared === 'text') return isScalar(actual)
  return false
}

export function isScalar(type: ValueType): boolean {
  return type === 'text' || type === 'number' || type === 'boolean' || type === 'datetime'
}

/**
 * Render a value as text at a `text` boundary.
 *
 * Numbers go through the ECMAScript `Number::toString` algorithm, which Go does
 * not implement — `String(1e-6)` is `"0.000001"` in JavaScript and `"1e-06"` in
 * Go. TypeScript gets it for free; the Go side ports it. A workflow that emails
 * a number must not read differently depending on which runner sent it.
 */
export function asText(value: Value): string {
  if (value === null) return ''
  if (typeof value === 'string') return value
  if (typeof value === 'number') return numberToText(value)
  if (typeof value === 'boolean') return value ? 'true' : 'false'
  if (value instanceof Date) return datetimeToText(value)
  return toJson(value)
}

/**
 * RFC 3339 in UTC, with a fractional part only when there is one.
 *
 * `Date.toISOString()` always writes three decimal places and Go's RFC3339Nano
 * writes none for a whole second, so neither language's default would do. This
 * spelling is Go's, and TypeScript is the one that has to be told.
 *
 * Instants carry millisecond precision, which is what `Date` can hold; the Go
 * side truncates to match rather than quietly keeping nanoseconds one runner
 * would print and the other could not.
 */
export function datetimeToText(value: Date): string {
  return value.toISOString().replace(/\.(\d*?)0*Z$/, (_whole, digits: string) =>
    digits === '' ? 'Z' : `.${digits}Z`,
  )
}

/**
 * Canonical JSON — what `json.stringify` produces.
 *
 * Hand-rolled in both languages rather than reaching for the built-in, because
 * the two built-ins disagree in ways that would reach a user: Go sorts object
 * keys and JavaScript preserves insertion order, Go escapes `<` and `&` by
 * default, and each formats numbers its own way. Sorting keys in both is the
 * only choice that can be made identical, so both sort.
 */
export function toJson(value: Value): string {
  if (value === null) return 'null'
  if (typeof value === 'string') return jsonString(value)
  if (typeof value === 'number') return numberToText(value)
  if (typeof value === 'boolean') return value ? 'true' : 'false'
  if (value instanceof Date) return jsonString(datetimeToText(value))
  if (Array.isArray(value)) return `[${value.map(toJson).join(',')}]`

  const entries = Object.keys(value)
    .sort()
    .map((key) => `${jsonString(key)}:${toJson((value as Record<string, Value>)[key] as Value)}`)
  return `{${entries.join(',')}}`
}

const JSON_ESCAPES: Record<string, string> = {
  '"': '\\"',
  '\\': '\\\\',
  '\n': '\\n',
  '\r': '\\r',
  '\t': '\\t',
  '\b': '\\b',
  '\f': '\\f',
}

function jsonString(value: string): string {
  const escaped = value.replace(/["\\\u0000-\u001f]/g, (char) => {
    const known = JSON_ESCAPES[char]
    if (known) return known
    return `\\u${char.charCodeAt(0).toString(16).padStart(4, '0')}`
  })
  return `"${escaped}"`
}

/** ECMAScript `Number::toString`. Named so the Go port has something to point at. */
export function numberToText(value: number): string {
  return String(value)
}
